nh3_testcubes.py

```diff
@@ -60,33 +112,75 @@ def generate_cubes(nCubes=100, nBorder=1, noise_rms=0.1,
-        # Find where centroids are too close
-        too_close = np.where(np.abs(Voff1-Voff2)<np.max(np.column_stack((Width1, Width2)), axis=1))
-        # Move the centroids farther apart by the length of largest line width 
-        min_Voff = np.min(np.column_stack((Voff2[too_close],Voff1[too_close])), axis=1)
-        max_Voff = np.max(np.column_stack((Voff2[too_close],Voff1[too_close])), axis=1)
-        Voff1[too_close]=min_Voff-np.max(np.column_stack((Width1[too_close], Width2[too_close])), axis=1)/2.
-        Voff2[too_close]=max_Voff+np.max(np.column_stack((Width1[too_close], Width2[too_close])), axis=1)/2.
+    Width1 = np.sqrt(Width1NT + 0.08 ** 2)
+    Width2 = np.sqrt(Width2NT + 0.08 ** 2)
 
-    scale = np.array([[0.2, 0.1, 0.5, 0.01]])
-    gradX1 = np.random.randn(nCubes, 4) * scale
-    gradY1 = np.random.randn(nCubes, 4) * scale
-    gradX2 = np.random.randn(nCubes, 4) * scale
-    gradY2 = np.random.randn(nCubes, 4) * scale
+    Temp = np.array([Temp1, Temp2]).swapaxes(0, 1)
+    Width = np.array([Width1, Width2]).swapaxes(0, 1)
+    Voff = np.array([Voff1, Voff2]).swapaxes(0, 1)
+    logN = np.array([logN1, logN2]).swapaxes(0, 1)
+
+    return nComps, Temp, Width, Voff, logN
+
+
+
+def make_cube(nComps, nBorder, xarr, Temp, Width, Voff, logN, gradX, gradY, noise_rms):
+    # the length of Temp, Width, Voff, logN, gradX, and gradY should match the number of components
+    xmat, ymat = np.indices((2 * nBorder + 1, 2 * nBorder + 1))
+    cube = np.zeros((xarr.shape[0], 2 * nBorder + 1, 2 * nBorder + 1))
+
+    results = {}
+    results['Tmax_a'], results['Tmax_b'] = (0,) * 2
+
+    for xx, yy in zip(xmat.flatten(), ymat.flatten()):
+
+        spec = np.zeros(cube.shape[0])
+
+        for j in range(nComps):
+            # define parameters
+            T = Temp[j] * (1 + gradX[j][0] * (xx - 1) + gradY[j][0] * (yy - 1)) + 5
+            if T < 2.74:
+                T = 2.74
+            W = np.abs(Width[j] * (1 + gradX[j][1] * (xx - 1) + gradY[j][1] * (yy - 1)))
+            V = Voff[j] + (gradX[j][2] * (xx - 1) + gradY[j][2] * (yy - 1))
+            N = logN[j] * (1 + gradX[j][3] * (xx - 1) + gradY[j][3] * (yy - 1))
+
+            # generate spectrum
+            spec_j = ammonia.cold_ammonia(xarr, T, ntot=N, width=W, xoff_v=V)
 
-    params1 = [{'ntot':14,
-                'width':1,
-                'xoff_v':0.0}] * nCubes
-    params2 = [{'ntot':14,
-                'width':1,
-                'xoff_v':0.0}] * nCubes
+            if (xx == nBorder) and (yy == nBorder):
+                Tmaxj = np.max(spec_j)
+                results['Tmax_{}'.format(ascii_lowercase[j])] = Tmaxj
+
+            # add each component to the total spectrum
+            spec = spec + spec_j
+
+        cube[:, yy, xx] = spec
+        if (xx == nBorder) and (yy == nBorder):
+            Tmax = np.max(spec)
+            results['Tmax'] = Tmax
+
+    # should we add noise independently on individual spectrum to avoid correlated noise across pixels?
```

Comment:
I'm not sure that I get this.  I think the cubes are set to be independent in the spatial direction (i.e., 1 beam width pixel spacing).

Reply:
Hmm, you're right on that one. I misunderstood line 164. 

I think I know what problem is now: there should be a - 0.5 term in that line. `numpy.random.rand` generates random samples from a uniform distribution over `[0, 1)` and not `[-0.5, 0.5)`.

Reply:
But it's `randn` not `rand`?